File: TST/tst_05.py
from flask_socketio import Namespace, emit
from com import Com

class tst_05(Namespace) :

    led_vals = {
        "all_off"            : 0 ,
        "vr_1"               : 1 ,
        "vr_2"               : 2 , 
        "vr_3"               : 4 ,
        "volt_limitter_high" : 8 ,
        "out_of_band_high"   : 10 ,
        "out_of_band_low"    : 20 ,
        "volt_limitter_low"  : 40 ,
        "auto_tap_blocked"   : 80 ,
        "reverse_power"      : 100 ,
        "volt_reduction"     : 200 ,
        "data_ports"         : 400 ,
        "super_visory_off"   : 800 ,
        "drag_hand_reset"    : 1_000 ,
        "alarm"              : 2_000 ,
        "warning"            : 4_000 ,
        "diag_error"         : 8_000 ,
        "ud_1"               : 10_000 ,
        "ud_2"               : 20_000 ,
        "ud_3"               : 40_000 ,
        "com_1_tx"           : 80_000 ,
        "com_1_rx"           : 100_000 ,
        "com_2_tx"           : 200_000 ,
        "com_2_rx"           : 400_000 ,
        "all_on"             : "7fffff"
    }
    cur_switched_on = ''
    succ_str = 'tst:05\r\n'
    
    # route..
    def on_led_status_cs(self, _id) :

        val = None
        if _id == tst_05.cur_switched_on :
            val = 0
        else :
            val = tst_05.led_vals[_id]

        
        res = tst_05.write_num(val)
        if res == True:
            tst_05.cur_switched_on = "all_off" if not val else _id 
            emit('led_status_sc',tst_05.cur_switched_on)
        else : 
            emit('error_sc',str(res))

    @staticmethod
    def write_num(num) :

        cmd_str = "TST:05 "
        num_str = str(num)
        cmd_str += "0"*(9 - len(num_str)) + num_str+'\n'
        Com.cur_input = cmd_str    
        Com.write_to_port(cmd_str.encode('ascii'))

        succ, data = Com.read_from_port()
        if (succ and tst_05.succ_str in data) :
            return True
        else :
            return data

File: static/TST/tst_05.js
// import {io} from '../libs/socketIO.js';
// import { io } from "socket.io-client";
import { io } from "https://cdn.socket.io/4.7.5/socket.io.esm.min.js";

export class tst_05 {
    constructor(url) {
        this.socket = io(url);   
        this.sender_init();
        this.receiver_init();
    }

    receiver_init() {
        // sc - server to client
        this.socket.on('led_status_sc', (_id) => {
            if (_id === "all_off"){
                this.prev_led.style.cssText =  '--border-top-color: white';
                this.id_ce.all_off.element.style.cssText = '--border-top-color: rgb(0, 255, 242)';
            } else {
                this.id_ce.all_off.element.style.cssText = '--border-top-color: white';
                this.prev_led.style.cssText =  '--border-top-color: white';
                this.id_ce[_id].element.style.cssText = '--border-top-color: ' + this.id_ce[_id].code;
                this.prev_led = this.id_ce[_id].element;
            }
        });
    }

    sender_init() {

        let color_code = {"orange" : "orange", "green" : "#b9e402", "red" : "red", "blue" : "rgb(0, 255, 242)"};

        // id color element
        this.id_ce = {
            "vr_1"              : {"code": color_code.orange, "element": null},
            "vr_2"              : {"code": color_code.orange, "element": null},
            "vr_3"              : {"code": color_code.orange, "element": null},
            "volt_limitter_high": {"code": color_code.orange, "element": null},
            "out_of_band_high"  : {"code": color_code.orange, "element": null},
            "out_of_band_low"   : {"code": color_code.orange, "element": null},
            "volt_limitter_low" : {"code": color_code.orange, "element": null},
            "auto_tap_blocked"  : {"code": color_code.orange, "element": null},
            "reverse_power"     : {"code": color_code.orange, "element": null},
            "volt_reduction"    : {"code": color_code.orange, "element": null},
            "ud_1"              : {"code": color_code.orange, "element": null},
            "ud_2"              : {"code": color_code.orange, "element": null},
            "ud_3"              : {"code": color_code.orange, "element": null},
            "warning"           : {"code": color_code.orange, "element": null},

            "data_ports"        : {"code": color_code.green,  "element": null},
            "com_1_rx"          : {"code": color_code.green,  "element": null},
            "com_2_rx"          : {"code": color_code.green,  "element": null},

            "alarm"             : {"code": color_code.red,    "element": null},
            "diag_error"        : {"code": color_code.red,    "element": null},
            "com_1_tx"          : {"code": color_code.red,    "element": null},
            "com_2_tx"          : {"code": color_code.red,    "element": null},
            "drag_hand_reset"   : {"code": color_code.red,    "element": null},
            "super_visory_off"  : {"code": color_code.red,    "element": null},

            "all_off"           : {"code": color_code.blue,   "element": null},
            "all_on"            : {"code": color_code.blue,   "element": null}
        }
        

        const _ids = Object.keys(this.id_ce);
        for (const _id of _ids){
            const element = document.getElementById(_id);
            element.socket = this.socket;
            this.id_ce[_id].element = element;
            this.id_ce[_id].element.addEventListener("click", function(){
                this.socket.emit('led_status_cs',_id);
            });
        }

        this.prev_led = this.id_ce.all_off.element;
    }

    // emit(_id){
    //     this.socket.emit('led_status_cs',_id)
    // }
}
